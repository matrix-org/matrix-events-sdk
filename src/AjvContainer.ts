/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import Ajv, {Schema, SchemaObject} from "ajv";
import AjvErrors from "ajv-errors";
import {NamespacedValue} from "./NamespacedValue";

export class AjvContainer {
    public static readonly ajv = new Ajv({
        allErrors: true,
    });

    static {
        AjvErrors(AjvContainer.ajv);
    }

    /* istanbul ignore next */
    // noinspection JSUnusedLocalSymbols
    private constructor() {}

    /**
     * Creates a JSON Schema representation of the EitherAnd<> TypeScript type.
     * @param ns The namespace to use in the EitherAnd<> type.
     * @param schema The schema to use as a value type for the namespace options.
     * @returns The EitherAnd<> type as a JSON Schema.
     */
    public static eitherAnd<S extends string = string, U extends string = string>(
        ns: NamespacedValue<S, U>,
        schema: Schema,
    ): {anyOf: SchemaObject[]; errorMessage: string} {
        // Dev note: ajv currently doesn't have a useful type for this stuff, but ideally it'd be smart enough to
        // have an "anyOf" type we can return.
        // Also note that we don't use oneOf: we manually construct it through a Type A, or Type B, or Type A+B list.
        if (!ns.altName) {
            throw new Error("Cannot create an EitherAnd<> JSON schema type without both stable and unstable values");
        }
        return {
            errorMessage: `schema does not apply to ${ns.stable} or ${ns.unstable}`,
            anyOf: [
                {
                    type: "object",
                    properties: {
                        [ns.name]: schema,
                    },
                    required: [ns.name],
                    errorMessage: {
                        properties: {
                            [ns.name]: `${ns.name} is required`,
                        },
                    },
                },
                {
                    type: "object",
                    properties: {
                        [ns.altName]: schema,
                    },
                    required: [ns.altName],
                    errorMessage: {
                        properties: {
                            [ns.altName]: `${ns.altName} is required`,
                        },
                    },
                },
                {
                    type: "object",
                    properties: {
                        [ns.name]: schema,
                        [ns.altName]: schema,
                    },
                    required: [ns.name, ns.altName],
                    errorMessage: {
                        properties: {
                            [ns.name]: `${ns.name} is required`,
                            [ns.altName]: `${ns.altName} is required`,
                        },
                    },
                },
            ],
        };
    }
}
