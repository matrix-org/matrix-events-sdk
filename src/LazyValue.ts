/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/**
 * Represents a lazily-loaded value. When accessed for the first time, the
 * value will be cached for the lifetime of the object. The getter function
 * will be released when the value is cached.
 */
export class LazyValue<T> {
    private cached: T | undefined;
    private getter: (() => T) | undefined;

    public constructor(getter: () => T) {
        // manually copy, so we don't expose the `undefined` type option
        this.getter = getter;
    }

    public get value(): T {
        if (this.getter !== undefined) {
            this.cached = this.getter();
            this.getter = undefined;
        }
        // Force a cast rather than assert because it's possible for `T` to
        // be undefined or null (we don't stop people from making bad decisions
        // in this class).
        return this.cached as T;
    }
}
