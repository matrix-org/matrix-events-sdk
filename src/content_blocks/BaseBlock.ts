/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import {ContentBlockWire} from "./types_wire";
import {InvalidBlockError} from "./InvalidBlockError";

/**
 * The simplest form of a content block in its parsed form.
 * @module Content Blocks
 */
export abstract class BaseBlock<T extends ContentBlockWire.Value> {
    private _raw: T | undefined = undefined;

    public get raw(): T {
        return this._raw!;
    }

    protected set raw(val: T) {
        if (val === undefined || val === null) {
            throw new InvalidBlockError(
                this.name,
                "Block value must be defined. Use a null-capable parser instead of passing such a value.",
            );
        }

        this._raw = val;
    }

    /**
     * Creates a new BaseBlock.
     * @param name The name of the block, for error messages and debugging.
     * @param raw The block's value.
     * @protected
     */
    protected constructor(public readonly name: string, raw: T) {
        this.raw = raw; // reuse validation logic
    }
}
