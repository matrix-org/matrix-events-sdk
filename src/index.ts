// created from ctix

// created from ctix

export * from "./content_blocks/m/EmoteBlock";
export * from "./content_blocks/m/MarkupBlock";
export * from "./content_blocks/m/NoticeBlock";
export * from "./events/m/EmoteEvent";
export * from "./events/m/MessageEvent";
export * from "./events/m/NoticeEvent";
export * from "./content_blocks/ArrayBlock";
export * from "./content_blocks/BaseBlock";
export * from "./content_blocks/BooleanBlock";
export * from "./content_blocks/IntegerBlock";
export * from "./content_blocks/InvalidBlockError";
export * from "./content_blocks/ObjectBlock";
export * from "./content_blocks/StringBlock";
export * from "./events/InvalidEventError";
export * from "./events/RoomEvent";
export * from "./LazyValue";
export * from "./NamespacedValue";
export * from "./types";
