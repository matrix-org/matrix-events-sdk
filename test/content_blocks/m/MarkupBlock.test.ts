/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import {MarkupBlock} from "../../../src";
import {testSharedContentBlockInputs} from "../BaseBlock.test";

describe("MarkupBlock", () => {
    testSharedContentBlockInputs("m.markup", [{body: "test"}], x => new MarkupBlock(x));

    it("should be able to identify the text and HTML representations", () => {
        const block = new MarkupBlock([
            {body: "neither", mimetype: "text/fail"},
            {body: "text here", mimetype: "text/plain"},
            {body: "html here", mimetype: "text/html"},
        ]);
        // noinspection DuplicatedCode
        expect(block.text).toStrictEqual("text here");
        expect(block.html).toStrictEqual("html here");
        expect(block.representationErrors).toBeDefined();
        expect(block.representationErrors.size).toStrictEqual(0);
        expect(block.representations).toBeDefined();
        expect(block.representations.length).toStrictEqual(3);
        expect(block.representations).toStrictEqual(block.raw);
    });

    it("should detect plain text as representations without mimetypes", () => {
        const block = new MarkupBlock([
            {body: "fail", mimetype: "text/fail"},
            {body: "text here"},
            {body: "html here", mimetype: "text/html"},
        ]);
        // noinspection DuplicatedCode
        expect(block.text).toStrictEqual("text here");
        expect(block.html).toStrictEqual("html here");
        expect(block.representationErrors).toBeDefined();
        expect(block.representationErrors.size).toStrictEqual(0);
        expect(block.representations).toBeDefined();
        expect(block.representations.length).toStrictEqual(3);
        expect(block.representations).toStrictEqual(block.raw);
    });

    it("should be able to handle missing text", () => {
        const block = new MarkupBlock([
            {body: "fail", mimetype: "text/fail"},
            {body: "html here", mimetype: "text/html"},
        ]);
        expect(block.text).toBeUndefined();
        expect(block.html).toStrictEqual("html here");
        expect(block.representationErrors).toBeDefined();
        expect(block.representationErrors.size).toStrictEqual(0);
        expect(block.representations).toBeDefined();
        expect(block.representations.length).toStrictEqual(2);
        expect(block.representations).toStrictEqual(block.raw);
    });

    it("should be able to handle missing HTML", () => {
        const block = new MarkupBlock([
            {body: "fail", mimetype: "text/fail"},
            {body: "text here", mimetype: "text/plain"},
        ]);
        expect(block.text).toStrictEqual("text here");
        expect(block.html).toBeUndefined();
        expect(block.representationErrors).toBeDefined();
        expect(block.representationErrors.size).toStrictEqual(0);
        expect(block.representations).toBeDefined();
        expect(block.representations.length).toStrictEqual(2);
        expect(block.representations).toStrictEqual(block.raw);
    });

    it("should be able to handle neither text or HTML", () => {
        const block = new MarkupBlock([{body: "nothing of use", mimetype: "text/fail"}]);
        expect(block.text).toBeUndefined();
        expect(block.html).toBeUndefined();
        expect(block.representationErrors).toBeDefined();
        expect(block.representationErrors.size).toStrictEqual(0);
        expect(block.representations).toBeDefined();
        expect(block.representations.length).toStrictEqual(1);
        expect(block.representations).toStrictEqual(block.raw);
    });

    it("should use the first text representation", () => {
        let block = new MarkupBlock([{body: "text here"}, {body: "not this text", mimetype: "text/plain"}]);
        // noinspection DuplicatedCode
        expect(block.text).toStrictEqual("text here");
        expect(block.representationErrors).toBeDefined();
        expect(block.representationErrors.size).toStrictEqual(0);
        expect(block.representations).toBeDefined();
        expect(block.representations.length).toStrictEqual(2);
        expect(block.representations).toStrictEqual(block.raw);

        // The same test, but inverting the undefined mimetype for safety
        block = new MarkupBlock([{body: "text here", mimetype: "text/plain"}, {body: "not this text"}]);
        // noinspection DuplicatedCode
        expect(block.text).toStrictEqual("text here");
        expect(block.representationErrors).toBeDefined();
        expect(block.representationErrors.size).toStrictEqual(0);
        expect(block.representations).toBeDefined();
        expect(block.representations.length).toStrictEqual(2);
        expect(block.representations).toStrictEqual(block.raw);
    });

    it("should handle having no representations provided", () => {
        const block = new MarkupBlock([]);
        expect(block.text).toBeUndefined();
        expect(block.html).toBeUndefined();
        expect(block.representationErrors).toBeDefined();
        expect(block.representationErrors.size).toStrictEqual(0);
        expect(block.representations).toBeDefined();
        expect(block.representations.length).toStrictEqual(0);
        expect(block.representations).toStrictEqual(block.raw);
    });

    it("should validate each representation has a defined body", () => {
        const block = new MarkupBlock([{mimetype: "text/plain"} as any]);
        // noinspection DuplicatedCode
        expect(block.text).toBeUndefined();
        expect(block.html).toBeUndefined();
        expect(block.representationErrors).toBeDefined();
        expect(block.representationErrors.size).toStrictEqual(1);
        expect(Array.from(block.representationErrors.entries())[0]).toMatchSnapshot();
        expect(block.representations).toBeDefined();
        expect(block.representations.length).toStrictEqual(0);
        expect(block.representations).toStrictEqual(block.raw);
    });

    it("should validate each representation has a valid body", () => {
        const block = new MarkupBlock([{body: true as any, mimetype: "text/plain"}]);
        // noinspection DuplicatedCode
        expect(block.text).toBeUndefined();
        expect(block.html).toBeUndefined();
        expect(block.representationErrors).toBeDefined();
        expect(block.representationErrors.size).toStrictEqual(1);
        expect(Array.from(block.representationErrors.entries())[0]).toMatchSnapshot();
        expect(block.representations).toBeDefined();
        expect(block.representations.length).toStrictEqual(0);
        expect(block.representations).toStrictEqual(block.raw);
    });

    it("should validate each representation has a valid mimetype", () => {
        const block = new MarkupBlock([{body: "text here", mimetype: true as any}]);
        // noinspection DuplicatedCode
        expect(block.text).toBeUndefined();
        expect(block.html).toBeUndefined();
        expect(block.representationErrors).toBeDefined();
        expect(block.representationErrors.size).toStrictEqual(1);
        expect(Array.from(block.representationErrors.entries())[0]).toMatchSnapshot();
        expect(block.representations).toBeDefined();
        expect(block.representations.length).toStrictEqual(0);
        expect(block.representations).toStrictEqual(block.raw);
    });

    it("should validate each representation is an object", () => {
        const block = new MarkupBlock([true as any]);
        // noinspection DuplicatedCode
        expect(block.text).toBeUndefined();
        expect(block.html).toBeUndefined();
        expect(block.representationErrors).toBeDefined();
        expect(block.representationErrors.size).toStrictEqual(1);
        expect(Array.from(block.representationErrors.entries())[0]).toMatchSnapshot();
        expect(block.representations).toBeDefined();
        expect(block.representations.length).toStrictEqual(0);
        expect(block.representations).toStrictEqual(block.raw);
    });

    it("should remove representations mid-array if they are invalid", () => {
        const block = new MarkupBlock([
            {body: "valid text"},
            {body: "more valid text", mimetype: "text/html"},
            {body: "text here", mimetype: true as any},
            {body: "even more text", mimetype: "text/plain"},
        ]);
        expect(block.text).toStrictEqual("valid text");
        expect(block.html).toStrictEqual("more valid text");
        expect(block.representationErrors).toBeDefined();
        expect(block.representationErrors.size).toStrictEqual(1);
        expect(Array.from(block.representationErrors.entries())[0]).toMatchSnapshot();
        expect(block.representations).toBeDefined();
        expect(block.representations.length).toStrictEqual(3);
        expect(block.representations).toStrictEqual(block.raw);
    });
});
