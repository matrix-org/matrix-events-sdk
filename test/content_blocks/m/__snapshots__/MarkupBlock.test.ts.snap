// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkupBlock should remove representations mid-array if they are invalid 1`] = `
[
  {
    "index": 2,
    "representation": {
      "body": "text here",
      "mimetype": true,
    },
  },
  [Error: m.markup[2]: mimetype should be a non-null string, or undefined (field not required)],
]
`;

exports[`MarkupBlock should validate each representation has a defined body 1`] = `
[
  {
    "index": 0,
    "representation": {
      "mimetype": "text/plain",
    },
  },
  [Error: m.markup[0]: must have required property 'body'],
]
`;

exports[`MarkupBlock should validate each representation has a valid body 1`] = `
[
  {
    "index": 0,
    "representation": {
      "body": true,
      "mimetype": "text/plain",
    },
  },
  [Error: m.markup[0]: body should be a non-null string and is required],
]
`;

exports[`MarkupBlock should validate each representation has a valid mimetype 1`] = `
[
  {
    "index": 0,
    "representation": {
      "body": "text here",
      "mimetype": true,
    },
  },
  [Error: m.markup[0]: mimetype should be a non-null string, or undefined (field not required)],
]
`;

exports[`MarkupBlock should validate each representation is an object 1`] = `
[
  {
    "index": 0,
    "representation": true,
  },
  [Error: m.markup[0]: must be object],
]
`;
